Add Viessmann

mpollmeier opened the pull request:
... mostly just looking for some feedback.
Based on the api investigation at https://github.com/evcc-io/evcc/issues/19639#issuecomment-2737462360

Fix https://github.com/evcc-io/evcc/issues/19639

### open questions
* why does `getmode` trigger two identical http requests?
* authentication via oauth2 using only the viessmann client id
* user guidance to figure out the installation id, gateway serial and device id
* should/can we configure a longer interval between updates? If so, how? Viessmann has an api limit of ~1 request per minute

### excerpt from my evcc.yaml
```
chargers:
  - name: viessmann
    type: template
    template: viessmann
    installation_id: 23XXXXX          # all digits
    gateway_serial: 76XXXXXXXXXXXXXX  # all digits
    device_id: 0
    target_temperature: 47
    token: eyJ.....
```


### get status
```
./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db charger viessmann
...
[viessmann] TRACE 2025/03/25 19:35:15 GET https://api.viessmann.com/iot/v2/features/installations/23XXXXX/gateways/76XXXXXXXXXXXXXX/devices/0/features/heating.dhw.oneTimeCharge
[viessmann] TRACE 2025/03/25 19:35:16 {
  "data": {
    ...
  }
}
[viessmann] TRACE 2025/03/25 19:35:16 GET https://api.viessmann.com/iot/v2/features/installations/23XXXXX/gateways/76XXXXXXXXXXXXXX/devices/0/features/heating.dhw.oneTimeCharge
[viessmann] TRACE 2025/03/25 19:35:16 {
  "data": {
    ...
  }
}
...
Charge status: B
Enabled:       false
Features:      [Heating IntegratedDevice]
```



### enable
```
./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db charger viessmann --enable
[viessmann] TRACE 2025/03/25 13:28:51 {"temperature": 47}
--
{
  "data": {
    "success": true,
    "message": null,
    "reason": "COMMAND_EXECUTION_SUCCESS"
  }
}
[viessmann] TRACE 2025/03/25 13:28:52 { }
--
{
  "data": {
    "success": true,
    "message": null,
    "reason": "COMMAND_EXECUTION_SUCCESS"
  }
}

```


### disable
```
./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db charger viessmann --disable
--
{
  "data": {
    "success": true,
    "message": null,
    "reason": "COMMAND_EXECUTION_SUCCESS"
  }
}
``

mpollmeier committed:
viessmann charger api: first draft for a template...

... mostly just looking for some feedback.
Based on the api investigation at https://github.com/evcc-io/evcc/issues/19639#issuecomment-2737462360

* why does `getmode` trigger two identical http requests?
* authentication via oauth2 using only the viessmann client id
* user guidance to figure out the installation id, gateway serial and device id
* should/can we configure a longer interval between updates? If so, how? Viessmann has an api limit of ~1 request per minute

```
chargers:
  - name: viessmann
    type: template
    template: viessmann
    installation_id: 23XXXXX          # all digits
    gateway_serial: 76XXXXXXXXXXXXXX  # all digits
    device_id: 0
    target_temperature: 47
    token: eyJ.....
```

```
./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db charger viessmann
...
[viessmann] TRACE 2025/03/25 19:35:15 GET https://api.viessmann.com/iot/v2/features/installations/23XXXXX/gateways/76XXXXXXXXXXXXXX/devices/0/features/heating.dhw.oneTimeCharge
[viessmann] TRACE 2025/03/25 19:35:16 {
  "data": {
    ...
  }
}
[viessmann] TRACE 2025/03/25 19:35:16 GET https://api.viessmann.com/iot/v2/features/installations/23XXXXX/gateways/76XXXXXXXXXXXXXX/devices/0/features/heating.dhw.oneTimeCharge
[viessmann] TRACE 2025/03/25 19:35:16 {
  "data": {
    ...
  }
}
...
Charge status: B
Enabled:       false
Features:      [Heating IntegratedDevice]
```

```
./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db charger viessmann --enable
[viessmann] TRACE 2025/03/25 13:28:51 {"temperature": 47}
--
{
  "data": {
    "success": true,
    "message": null,
    "reason": "COMMAND_EXECUTION_SUCCESS"
  }
}
[viessmann] TRACE 2025/03/25 13:28:52 { }
--
{
  "data": {
    "success": true,
    "message": null,
    "reason": "COMMAND_EXECUTION_SUCCESS"
  }
}

```

```
./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db charger viessmann --disable
--
{
  "data": {
    "success": true,
    "message": null,
    "reason": "COMMAND_EXECUTION_SUCCESS"
  }
}
```

andig commented:
Sieht gut aus 👍🏻

mpollmeier commented:
Danke! Hast du evtl Ideen / Anmerkungen zu den 'open questions'?

andig commented:
> why does getmode trigger two identical http requests?

It's two values and there are no cache headers. Add `cache: 1s` to force caching.

> authentication via oauth2 using only the viessmann client id

We still need the non-browser flow for that, then I have an implementation idea.

> user guidance to figure out the installation id, gateway serial and device id

up to you :)

> should/can we configure a longer interval between updates? 

ATM only increasing global interval.

mpollmeier commented:
Got the browserless authentication working with shell/curl/jq, based on https://documentation.viessmann.com/static/authentication

### Prerequisite (only once)
The user must manually log into the [Viessmann Developer Portal](https://app.developer.viessmann.com/) and create a client: disable google captcha and configure any redirect URI (the default value `http://localhost:4200/` is fine):

<img src="https://github.com/user-attachments/assets/bc2d47f7-3027-4d78-b3b4-e424fff743d4" width="400" />


### Create an access token with two http requests
```
# from the viessmann developer portal (all digits)
VIESSMANN_CLIENT_ID=55f2c71XXXXXXXXXXXXXXXXXXXXXXXXX

# must equal the value configured in viessmann developer portal for this client id
VIESSMANN_REDIRECT_URI="http://localhost:4200/"

# can be static, see notes below
VIESSMANN_CODE_CHALLENGE="5M5nhkBfkWZCGfLZYcTL-l7esjPUN7PpZ4rq8k4cmys"
VIESSMANN_CODE_VERIFIER="6PygdmeK8JKPuuftlkc6q4ceyvjhMM_a_cJrPbcmcLc-SPjx2ZXTYr-SOofPUBydQ3McNYRy7Hibc2L2WtVLJFpOQ~Qbgic455ArKjUz9_UiTLnO6q8A3e.I_fIF8hAo"

# user credentials for the viessmann developer portal
VIESSMANN_USER=michael@michaelpollmeier.com
 VIESSMANN_PASS='secret'
 
VIESSMANN_CODE=$(curl -X POST --silent \
  --user $VIESSMANN_USER:$VIESSMANN_PASS \
  --output /dev/null \
  --dump-header -    \
  "https://iam.viessmann.com/idp/v3/authorize?client_id=$VIESSMANN_CLIENT_ID&redirect_uri=$VIESSMANN_REDIRECT_URI&scope=IoT%20User%20offline_access&response_type=code&code_challenge=$VIESSMANN_CODE_CHALLENGE&code_challenge_method=S256" \
  | grep "^location: "            \
  | sed 's/.*\?code=\(.*\).*/\1/' \
  | tr -d '[:space:]')
  
# create access token
TOKEN_RESPONSE=$(curl -XPOST --silent \
  -H "Content-Type: application/x-www-form-urlencoded" \
  --data "grant_type=authorization_code&client_id=$VIESSMANN_CLIENT_ID&redirect_uri=$VIESSMANN_REDIRECT_URI&code_verifier=$VIESSMANN_CODE_VERIFIER&code=$VIESSMANN_CODE" \
  https://iam.viessmann.com/idp/v3/token)
echo $TOKEN_RESPONSE
VIESSMANN_TOKEN=$(echo $TOKEN_RESPONSE | jq --raw-output .access_token)
VIESSMANN_REFRESH_TOKEN=$(echo $TOKEN_RESPONSE | jq --raw-output .refresh_token)
```

### Notes
* the values for code_challenge and code_verifier are taken straight from https://documentation.viessmann.com/static/authentication - in theory we can create new ones for each request, but I'm not sure if we'd really have any benefit
* the code value is in the response header `location` in the first
* trimming whitespace is important, at least for the above shell commands - the code has length 43, check e.g. via `echo ${#VIESSMANN_CODE}`
* the code that we get in the first request is only valid for 20 seconds - afterwards Viessman responds with an 'invalid grant' error
* the access token expires after 1 hour (3600s)

### Token TTL and refresh token
We can generate a new token with the `VIESSMANN_REFRESH_TOKEN`, and according to the [documentation](https://documentation.viessmann.com/static/authentication) that should have a longer TTL of 180 days, but when I tried the TTL was still always 3600s for every newly generated or refreshed token :(
To keep things simple it probably makes sense to generate a new token, store it for one hour, and then start from scratch. 
Just in case: here's how to get a new token based on the REFRESH_TOKEN we got previously:
```
curl -X POST "https://iam.viessmann.com/idp/v3/token" \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -d "grant_type=refresh_token&client_id=$VIESSMANN_CLIENT_ID&refresh_token=$VIESSMANN_REFRESH_TOKEN"
```

mpollmeier committed:
configure cache to prevent two identical requests within 1 second

andig commented:
Also OAuth2 code exchange. Da kann ich mir was einfallen lassen, dauert aber ein bisschen.

mpollmeier commented:
Genau. Ich hab auch den Teil über PKCE nochmal korrigiert, da sollten wir wirklich neue code challenge und verifier generieren für jeden request. 

Das PKCE handling hab ich mir gerade mal von ner KI schreiben lassen, das funktioniert auch:
```bash
go mod init go-auth2-test
go get golang.org/x/oauth2
```

main.go:
```
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"golang.org/x/oauth2"
)

// GeneratePKCE creates a new OAuth2 code verifier and challenge using S256.
func GeneratePKCE() (oauth2.AuthCodeOption, oauth2.AuthCodeOption, string) {
	// Generate a random 32-byte code verifier
	verifierBytes := make([]byte, 32)
	rand.Read(verifierBytes)
	codeVerifier := base64.RawURLEncoding.EncodeToString(verifierBytes)

	// Hash the verifier using SHA256 and encode as base64 URL
	hash := sha256.Sum256([]byte(codeVerifier))
	codeChallenge := base64.RawURLEncoding.EncodeToString(hash[:])

	// Return options for OAuth2 and the code verifier
	return oauth2.SetAuthURLParam("code_challenge", codeChallenge),
		oauth2.SetAuthURLParam("code_challenge_method", "S256"),
		codeVerifier
}

func main() {
	codeChallengeOption, codeChallengeMethodOption, codeVerifier := GeneratePKCE()

	fmt.Println("PKCE Code Verifier:", codeVerifier)
	fmt.Println("PKCE Code Challenge:", codeChallengeOption)
	fmt.Println("PKCE Code Challenge Method:", codeChallengeMethodOption)
}
```

```bash
go run main.go
PKCE Code Verifier: NUBS81DWRCNOKHqrQvWZpqr2f7VLwiIHcY4A2jbodDI
PKCE Code Challenge: {code_challenge Zps-JR5UfvgOHTk7LeClyM7-zFydhdbRO6uwQMolUnk}
PKCE Code Challenge Method: {code_challenge_method S256}
```

Aber wie man das token handling / renewal am besten macht und wo genau man es in evcc unterbringt, ist vermutlich noch die schwierigere Frage, da kennst du dich besser aus.

andig commented:
Brauchst du alles nicht- pkce ist bei Go schon enthalten. Meine Idee wäre, https://github.com/evcc-io/evcc/pull/20066 so zu erweitern, dass wir dort neben basic/ bearer auch Authorisation Plugins einführen die ihrerseits einen autorisierten `http.Transport` erzeugen. Damit wird das  `Auth` Objekt wichtiger. Willst Du es mal probieren?

/cc @thecem wir hatten das Thema noch bei einer weiteren WP.

thecem commented:
https://github.com/evcc-io/evcc/discussions/19904

andig commented:
I've updated https://github.com/evcc-io/evcc/pull/20066 to add the required infrastructure.

andig commented:
Z.B. https://github.com/evcc-io/evcc/blob/master/vehicle/porsche/identity.go wäre ein gutes Beispiel für eine `oauth2.TokenSource`. Diese dann in einen `http.Transport` zu verwandeln ist trivial:

```go
var tokenSource oauth2.TokenSource

tokenSource = ...

transport := &oauth2.Transport{
    Source: tokenSource,
    Base: base,
}
```

andig commented:
Hab ich ergänzt, vmtl. ist das mit den User noch falsch (das muss ein basic auth header sein?).

mpollmeier committed:
WIP - getting the code from viessmann works :tada:

mpollmeier commented:
@andig ich hab mich heute mal damit beschäftigt und werde (langsam) mit `go` warm.

Den `CODE`, also den ersten request, bekomme ich jetzt schonmal. Da konnte ich tatsächlich viel von der `porsche/identity.go` wiederverwenden. 

Ich teste lokal mit `go run viessmann-test.go` und https://github.com/mpollmeier/evcc/commit/db223cd81c42f46e8fdf94cfaf89a5b1fa2b7e2e

Wenn du Zeit für early feedback hast: gerne. Ansonsten wollte ich nur mal ein Lebenszeichen senden, bzw die Info, dass ich dran bin. Ist halt auch immer von der verfügbaren Freizeit abhängig.

andig commented on charger/viessmann/identity.go:
```diff
@@ -0,0 +1,170 @@
+import (
+	"context"
+	"fmt"
+  "os"
```
Du brauchst gofmt oder einfach VScode

andig replied:
Die Datei selbst muss nach plugin/auth

mpollmeier replied:
Ich verstehe nicht ganz - soll die generisch werden und als `plugin/auth/identity.go` dort liegen, in der Hoffnung dass alle oauth2 provider identisch sind? Oder Viessmann-spezifisch als `plugin/auth/viessmann.go`?

andig replied:
Letzteres. Deshalb muss sie sich mit eigenem Namen registrieren.

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,65 @@
+    cache: 2s # to prevent making two identical requests straight after each other for "getmode"
+    headers:  
+      - content-type: application/json
+      - authorization: Bearer {{.token}}
```
Siehe https://github.com/evcc-io/evcc/pull/20066

mpollmeier committed:
requesting a token works now. not sure about refresh token though

mpollmeier commented:
Aktueller Status auf diesem Branch: code und token abrufen klappt. Refresh token hab ich mit der go api nicht hinbekommen, vielleicht kannst du mir auf die Sprünge helfen, dann kann ich das testen.
```
go run viessmann-test.go

&{eyJraWQiOiI0ZWNhM2V....UjaN8NT5OJ-QcZA7LpNqec expires_in:3600 refresh_token:1d51abffeb973fd8168b0869e0f3976b token_type:Bearer] 0}
```

hbpv commented:
Hey. So weit ich das sehe, kann ich nur einen Client im ViessmannDevPortal einrichten. Ich nutze diesen mit einer Config für die Viessmann HA Integration. Dort habe ich als URI „vicare://oauth-callback/everest“

Ich kann da jetzt noch eine hinzufügen, und die selbe Client ID nutzen.. aber ob das funktioniert?

![IMG_8623](https://github.com/user-attachments/assets/845a7f3a-302e-4363-bda3-18e6a6ff5670)

was aber dann auf jeden Fall für User wie mich, die die Integration zur Steuerung/Statis Info nutzen, ein Problem wird, sind die limitierten API Calls im kostenlosen Account. Die reichen im Monat für alle Entitäten einer WP aus, aber kommt dort noch evcc hinzu, wird das nicht reichen und den recht teuren pro Account erfordern..

aber trotzdem: geile Arbeit!

mpollmeier commented:
@hbpv die redirect URI ist für unseren use case egal, es muss nur derselbe sein für die beiden requests die wir machen. Das kannst du also auch auf dem "everest" lassen. Man kann also denselben client benutzen, wenn man schon einen anderen use case hat. 

Bzgl. API call limitation: man darf 1450 api calls pro 24h mit dem kostenlosen Account machen, also einen call pro Minute. Wenn wir das mit dem caching vernünftig konfigurieren und vielleicht noch ein bisschen bewusste Trägheit einbauen, sehe ich hier überhaupt kein Problem. 
Wir wollen und brauchen ja nicht ständig die aktuelle Temperatur auslesen, das reicht auch alle 20min oder so. Und genauso wollen wir ja die Warmwasserbereitung nicht alle zwei minuten an- und wieder ausschalten, sondern auch da eine gewisse Trägheit. Ich hab noch nicht danach gesucht, vermute aber, dass evcc das so schon bei anderen WP und Chargern so macht. 

https://developer.viessmann.com/start/faq.html?appState=API%252Busage%253Dmaximum%252Bapi%252Bcalls

mpollmeier committed:
explicitly invoke TokenWithExpiry - doesn't seem to make a difference though

```
go run viessmann-test.go

&{eyJraWQiOiI0Z..... expires_in:3600 refresh_token:bae936510898c1659b03c0baee57830d token_type:Bearer] 0}
2025-04-13 14:59:41.637962037 +0200 CEST m=+3601.165162594
```

mpollmeier commented:
@andig ich rufe jetzt `util.TokenWithExpiry` explizit auf. Das schadet nicht, aber ich denke es wäre gar nicht nötig - es passiert ja nichts, weil `token.ExpiresIn` bereits gesetzt ist. 
Siehe https://github.com/evcc-io/evcc/blob/1eafc455cfe535d9d35f0ed83afd383876154008/util/token.go#L10

andig commented:
Das ist wie gesagt ein Trugschluss.

andig commented:
Nächster Schritt wäre jetzt Deine identity   in einen auth.Transport zu verwandeln und ins Template zu integrieren.

mpollmeier commented:
> Nächster Schritt wäre jetzt Deine identity in einen auth.Transport zu verwandeln und ins Template zu integrieren.

Daran habe ich mich eben versucht, aber meine go Skills kommen da an ihre Grenzen.

1) verstehe ich die go package und file Struktur noch nicht, und nichts funktionierte nicht mehr, sobald ich die Datei in plugin verschoben habe (und das package angepasst habe). 

2) verstehe ich nicht warum wir ein http.Transport brauchen bzw. wie ich das umsetzen soll. Ich spreche doch mit einer OAuth API, da ist doch http ein level zu tief. Oder bleibt die OAuth2Config.Exchange wie gehabt, und ein http.Transport wird nochmal drum gewrapped?

mpollmeier commented:
Btw du darfst auch gern einfach auf diesen Branch pushen...
![image](https://github.com/user-attachments/assets/97f2d899-eae6-4c55-a748-09a869725208)
https://github.com/mpollmeier/evcc/tree/charger/viessmann

andig commented:
> 2. verstehe ich nicht warum wir ein http.Transport brauchen bzw. wie ich das umsetzen soll. Ich spreche doch mit einer OAuth API, da ist doch http ein level zu tief. Oder bleibt die OAuth2Config.Exchange wie gehabt, und ein http.Transport wird nochmal drum gewrapped?

Den brauchst du damit die HTTP Aufrufe aus dem Template das Token nutzen.

andig committed:
Convert identity into auth plugin

andig commented:
1. ich hab aus der identity jetzt ein auth plugin gemacht- das musst du noch in das template (bei jedem request!) integrieren
2. dein code hat bisher das refreshtoken nicht genutzt sondern jedesmal ein login gemacht. Ich habs auf refreshtoken umgebaut. ob das so geht und Vaillant ein refreshtoken raus rückt weiss ich natürlich nicht

Ohe Zugangsdaten kann ich da auch nix mehr testen ;)

andig committed:
wip

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,65 @@
+  - name: target_temperature
+    description:
+      de: Zieltemperatur für Einmal-Warmwasser-Zubereitung (°C), max=60
+      en: Target temperator for onetimecharge (°C), max=60
```
```suggestion
      en: Target temperature for one-time charge (°C), max=60
```

andig committed:
Use template vars for auth

andig committed:
Update templates/definition/charger/viessmann.yaml

andig committed:
Fix build

andig committed:
No test

andig committed:
No helper

mpollmeier committed:
mark params as required and add descriptions

mpollmeier commented:
So schauts aktuell in der UI aus, ich hab die Labels angepasst und die Parameter als required markiert. Die Authentifizierung und Statusabfrage haben auch schon grundsätzlich funktioniert, ich muss aber morgen mal weitertesten wenn mein API Guthaben wieder aufgefüllt ist. Insb. würde mich interessieren, ob das RefreshToken Handling funktioniert. 

![image](https://github.com/user-attachments/assets/1007b885-fd53-4269-92df-f9523c1ecfa5)

andig commented:
Du kannst

    evcc charger

benutzen und damit nach Herzenslust schalten.

MasterGoofy commented:
Hallo,,Vielen Dank schon einmal für euren Einsatz. 👍Wie kann ich das ganze denn bei mir testen? Muss es dazu erst in den offiziellen Build übernommen werden?Viele Grüße DanielVon: andig ***@***.***>Gesendet: Dienstag, April 15, 2025 8:37:17 PMAn: evcc-io/evcc ***@***.***>Cc: MasterGoofy ***@***.***>; Manual ***@***.***>Betreff: Re: [evcc-io/evcc] Add Viessmann (PR #20111)


andig left a comment (evcc-io/evcc#20111)Du kannstevcc charger
benutzen und damit nach Herzenslust schalten.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.

andig commented:
Du musst selbst compilieren

mpollmeier commented:
Genau, also diesen branch auschecken, also entweder mit `git remote add` wenn du evcc schon lokal in git hast, oder frisch:
```
git clone git@github.com:mpollmeier/evcc.git
cd evcc
git checkout charger/viessmann

make install-ui
make install
make

./evcc
```
siehe CONTRIBUTING.md

andig commented:
@mpollmeier was fehlt denn noch? Hast du es schon getestet?

mpollmeier commented:
Token generieren und Status abfragen funktionieren. 
Beim Temperatur-einstellen-request fehlt wohl noch ein http header, das sollte ich aber selbst hinkriegen. 

Wo ich deine Hilfe brauche @andig: der Token wird aber für jede Anfrage neu abgerufen, sowohl wenn ich evcc als Service laufen lasse (1) und dann in der UI auf `validate` gehe, als auch wenn ich `charger` (2) aufrufe. Sollten nicht in beiden Fällen (mindestens aber im ersten) der Token wiederverwendet werden? Ich sehe im output jeweils die http requests um ein neues Token von Viessman zu holen. 

Hast du ne Idee woran das liegen könnte? Wenn ich irgendwas testen soll, kannst du auch gern ein `println` einbauen und ich lasse es laufen. Oder du erstellst dir selbst nen Viessmann Account zum testen. 

(1) `./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db`
(2) `./evcc --log trace -c ~/.evcc/evcc.yaml --database ~/.evcc/evcc.db charger`

andig commented:
Magst mal info@evcc.io Zugangsdaten zum testen schicken? Dann kann ichs mir anschauen.

mpollmeier commented:
Hab nen account für info@evcc.io registriert, bitte klick auf den Bestätigungslink aus der email. Das Passwort habe ich gerade an cpuidle@gmx.de geschickt. Du musst dir dann noch im Developer Portal eine Client ID generieren: https://app.developer.viessmann.com

andig commented:
Danke, aber dafür fehlt mir die Zeit wie auch das Gerät...

mpollmeier commented:
Dann hatte ich dich falsch verstanden - wie meintest du das denn mit "Zugangsdaten zum testen"?

Ein Gerät brauchst du ja gar nicht, das Problem ist beim OAuth, dass evcc nicht den vorhandenen und validen Token verwendet, sondern sich immer wieder einen neuen holt.

andig commented:
Wie soll ich das denn testen? Das passiert ja immer nur bei einem API-Request?

mpollmeier commented:
Zunächst wie oben beschrieben, also:
* den Account den ich dir erstellt habe für info@evcc.io einmal Bestätigen (link in email klicken)
* password ist in deiner gmx email
* clientId musst du kurz erstellen im dev portal

Dann in evcc einen viessmann charger hinzufügen, für InstallationID, Gateway Serial und Device ID trägst du jeweils 0 ein. Dann auf `validate` klicken, dann holt sich evcc einen OAuth Token. Bei der validierung dann von Viessman ein 404 zurück, aber das interessiert erstmal nicht. Dann nochmal `validate`, und evcc sollte evcc dann eigentlich den noch validen token benutzen. Stattdessen holt er sich aber aktuell immer einen neuen...

andig commented:
Das ist erwartet. Nach validate wird das ganze Objekt wieder abgeräumt.

mpollmeier committed:
set http header for json content-type

mpollmeier commented:
Den requests zum enable/disable etc. habe ich noch nen header für `content-type: application/json` hinzugefügt, diese funktionieren jetzt. 

Das token handling funktioniert im "Normalbetrieb" wunderbar - der Token wird eine Stunde lang verwendet, und dann wird ein neuer abgerufen. 

Wir sollen in der Dokumentation darauf hinweisen, dass das `interval` auf mindestens 60s (besser noch etwas höher) eingestellt werden sollte, weil Viessman ein API limit von ca 1 request/minute hat. Und es wäre gut, wenn man das in Zukunft pro device einstellen kann, nicht nur global. 

Fehlt aus deiner Sicht sonst noch irgendwas?

andig commented:
Wirklich schön wäre ja jetzt, wenn wir die GetMode Abfrage cachen könnten, müssten sie bei SetMode aber resetten. Dazu fällt mir leider kein guter Mechanismus fürs Template ein :/

andig commented on plugin/auth/viessmann.go:
```diff
@@ -0,0 +1,127 @@
+
+	return oc.Exchange(ctx, code, oauth2.VerifierOption(cv))
+}
+
+// func (v *Viessmann) RefreshToken(token *oauth2.Token) (*oauth2.Token, error) {
+// 	ctx := context.WithValue(context.Background(), oauth2.HTTPClient, v.Client)
+// 	ts := oauth2.ReuseTokenSource(token, OAuth2Config.TokenSource(ctx, token))
+
+// 	token, err := ts.Token()
+// 	if err != nil {
+// 		token, err = v.login("x", "x")
+// 	}
+
+// 	return util.TokenWithExpiry(token), err
+// }
```
```suggestion
```

andig replied:
```suggestion
```

mpollmeier commented:
Der Gedanke ist richtig, aber wie immer ist das mit Caches so eine Sache. Der User kann diese Einstellungen halt auch manuell in der Viesmann App (ViCare) setzen. Trotzdem wäre es gut das zu haben, eir könnten da ja in der Dokumentation drauf hinweisen.

mpollmeier commented:
Wie schauts aus, fehlt noch was oder magst du mergen?

andig commented:
Du hast noch eine längere Liste an TODOs im Tempalte dokumentiert

mpollmeier committed:
add help - requires updated docs

mpollmeier committed:
fine tune

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,118 @@
+      de: InstallationID
+      en: InstallationID
+    help:
+      de: https://docs.evcc.io/docs/devices/heating#viessmann
```
Das sieht aus wir ein Zirkelbezug- was würde ich als Anwender mit dem Link bzgl. InstallationID machen?

mpollmeier replied:
Ich schreibe mal die Dokumentation dafür. Das ist zu kompliziert für diese Stelle, daher der link...

andig replied:
So funktioniert das nicht- das Template _ist_ die Doku!

mpollmeier replied:
Ich sehe gerade erst: der content für die Dokumentation wird ja aus dem Template generiert, d.h. ich kann bzw sollte es direkt in das Template schreiben, und nicht in das docs repo.

mpollmeier replied:
Problem: die Installation ID ist leider nicht in der Viessmann App einsehbar, diese bekommt man offenbar nur über die API. Ich hatte das ja bei mir via `curl` und `jq` gemacht, aber das sind insgesamt 3-4 requests (weil man ja auch erstmal oauth machen muss etc.), daher ist das für den Anwender vermutlich zu kompliziert, und nicht jeder Anwender hat diese Tools. 

Da evcc ja schon bei der Einrichtung bzw. Validierung mit der Viessmann API kommuniziert: können wir an der Stelle den Wert der Installation ID auslesen und speichern? So funktioniert es auf Kommandozeile:
```
curl -X GET https://api.viessmann.com/iot/v1/equipment/installations \
  -H "Authorization: Bearer $VIESSMANN_TOKEN" | jq '.data[].id'
```

Darin liegt aber auch eine Chance: falls das geht, können wir dasselbe mit der Gateway Serial machen, der Anwender muss diese dann nicht manuell aus der Viessmann App raussuchen. Ähnliches gilt für die Device ID.

mpollmeier replied:
Bei Vaillant wird das schon so gemacht, ich vermute das funktioniert nicht mit unserem Template Ansatz...?
https://github.com/evcc-io/evcc/blob/95d70cf4345550c255cd03d4ec20e12e3cc21ca2/charger/vaillant.go#L93

andig replied:
Ne, das klappt leider nicht. Da bliebe nur eine gute Doku für den Wert zu schreiben. Vielleicht könnte man sich auch noch irgendwas mit Go Plugin ausdenken das den Wert liefert und cached, spontan fällt mir aber nichts ein. Bleibt also (erstmal) nur Doku :(

mpollmeier replied:
Ok. Wo wäre der richtige Ort dafür? Im docs repo vermutlich, in einem Bereich der nicht generiert wird?

andig replied:
Am besten hier im Template- dort wird's gebraucht.

mpollmeier replied:
Some lines are so long (specifically the one with `VIESSMANN_CODE_VERIFIER`) that they break the entire layout... screenshots below. Any ideas on how to fix that, or should we go the initial route of adding the docs elsewhere and just linking them here?

Btw I tried to use `<code>` blocks instead of ```, but that doesn't help, because then it interprets some characters as font styles, e.g. the `_` isn't displayed and instead it renders the following characters in bold text :(


![Screenshot_2025-05-19_17-30-58](https://github.com/user-attachments/assets/5652335d-2dcc-403c-a8e3-e356ee0304aa)


![image](https://github.com/user-attachments/assets/336cddaf-016d-4382-8fdc-ca8a32ecb1f2)

andig replied:
@naltatis ?

mpollmeier replied:
vielleicht doch besser ins doc repo?

naltatis replied:
Ich hab das Rendering angepasst, sodass der Code-Block nicht mehr das Layout sprengt. Generell finde ich den Ort und auch die Formatierung als Code schon gut.

<img width="567" alt="Bildschirmfoto 2025-05-29 um 15 43 35" src="https://github.com/user-attachments/assets/3a2d8f9a-0d39-46b7-b836-b89710a6e1af" />

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,118 @@
+      de: Device ID
+      en: Device ID
+    help:
+      de: https://docs.evcc.io/docs/devices/heating#viessmann
```
Dito

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,118 @@
+          clientid: {{ .clientid }}
+        body: >
+          { }
+    - case: 2 # boost
```
Bei Vaillant haben wir das so geändert, dass WW alle 15min re-boosted wird da das sonst nur 1x der Fall ist. Das ließe sich per watchdog machen.

mpollmeier replied:
Ich dachte bisher, die Anlage heizt der solange bis die Solltemperatur erreicht ist, aber es schadet vermutlich nicht, den watchdog einzuschalten. Wie funktioniert das?
In der Statusabfrage bekommen wir auch die Info, ob WW Booster an oder aus ist...

mpollmeier replied:
Ich habe gerade mal manuell mit der Viessmann App getestet: WW bleibt solange aktiv, bis die Solltemperatur erreicht ist. Watchdog sollte also nicht nötig sein.

andig replied:
> bis die Solltemperatur erreicht ist. 

Und dann? Kühlt der Puffer wieder aus...

mpollmeier replied:
Wenn die Zieltemperatur erreicht ist oder wir die WW Erzeugung ausschalten (z.b. weil kein PV Überschuss mehr vorhanden ist), wechselt die Anlage wieder in den Normalbetrieb. Genau das wollen wir doch..?

andig replied:
Und wenn weiter Überschuss da ist?

MasterGoofy replied:
Könnte man dafür nicht eine Untergrenze der Temperatur angeben? Beispiel : 
WW hat 60 Grad erreicht und Überschuss ist weiterhin vorhanden. Jetzt macht es keinen Sinn direkt wieder die Wärmepumpe zu starten. Stattdessen erst wieder wenn das Wasser auf eine einstellbare Temperatur gefallen ist.

andig replied:
Nur wenn du eine Idee hast wie man das im Template ausdrücken wollte. Warum nicht einfach boosten? Falls der Speicher noch warm ist passiert ja nix?

mpollmeier replied:
Alle 30min hört sich gut an, das können wir gerne machen. Die Vaillant Implementierung hast du ja in https://github.com/evcc-io/evcc/pull/20752 gemacht. Das ist allerdings in `charger/vaillant.go`, für viessmann haben wir aktuell nur `auth/viessmann.go` und benutzen ansonsten das Template. Kannst du das umstellen bzw einbauen?

andig replied:
Du kannst einfach das watchdog plugin mit timeout: 60m nehmen? Reset wenn wieder auf normal gestellt wird.

mpollmeier replied:
Ich hab mal was gepushed, magst du mal schauen?
Ist von 4d9b54288720de9cd4d982eb307aeed1dc35bf0d abgekupfert...

mpollmeier committed:
reword help texts

mpollmeier committed:
description

mpollmeier committed:
device id is typically 0

mpollmeier committed:
add watchdog - retrigger every 30mins

andig committed:
Update plugin/auth/viessmann.go

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -75,40 +82,15 @@ render: |
-      set:
-        source: sequence
+    source: watchdog
+    timeout: {{ .watchdog }} # re-write at timeout/2
```
This needs to reset on mode 1

mpollmeier replied:
Oh nvm, I think I just found it. Can you take a look please?

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -60,6 +60,13 @@ params:
     required: true
     default: 45
     type: int
+  - name: watchdog
```
This isn't particularly clear. Trigger what? Why don't we just hard-code a 30m interval below?

mpollmeier replied:
I just though if it's under advanced it might be useful to some users. 
But I don't mind, just changed this to hard coded - does it look right?
https://github.com/evcc-io/evcc/pull/20111/commits/1ec099e3475b821cf7e87c6221857224e874f91d

mpollmeier committed:
hardcode watchdog timeout to 60min / 30min

mpollmeier committed:
reset watchdog on mode 1 (normal)

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,125 @@
+      de: InstallationID
+      en: InstallationID
+    help:
+      de: siehe https://docs.evcc.io/docs/devices/heating#viessmann
```
Das geht immer noch nicht ;)

mpollmeier replied:
Das wartet noch auf deine Antwort auf https://github.com/evcc-io/evcc/pull/20111#discussion_r2072399041 / https://github.com/evcc-io/evcc/pull/20111#discussion_r2072609314

Es wäre gut bzw wichtig, wenn wir die InstallationID beim hinzufügen des Geräts einmalig via Viessman API abholen könnten. Falls wir das nicht hinbekommen, schließen wir alle user aus, die kein `curl` / `jq` etc. haben bzw bedienen können. :(

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,125 @@
+  - name: gateway_serial
+    required: true
+    description:
+      de: Gateway Serial
```
```suggestion
      de: Gateway Serialnummer
```

andig commented on templates/definition/charger/viessmann.yaml:
```diff
@@ -0,0 +1,125 @@
+    # true -> oneTimeCharge is enabled -> boost mode -> 2
+  setmode:
+    source: watchdog
+    timeout: 60m # re-write at timeout/2
```
```suggestion
    timeout: 30m # re-write at timeout/2
```

mpollmeier committed:
helptext WIP - layout is broken...

see https://github.com/evcc-io/evcc/pull/20111#discussion_r2096212879

naltatis committed:
code formatting; template improvements

naltatis committed:
Merge branch 'master' into charger/viessmann

andig commented on templates/definition/charger/viessmann.yaml:
````diff
@@ -0,0 +1,148 @@
+        VIESSMANN_CLIENT_ID=<your-clientid>
+        ```
+
+        Then execute the following (n.b. it's best to paste the entire block as-is, since the intermediate 'CODE' is only valid for 20 seconds):
````
Ich hab mir die Doku nochmal angeschaut. Dieser Teil ergibt keinen Sinn. Der `curl` Aufruf fehlt!

MasterGoofy commented:
Da es die Integration nun ins offizielle Build geschafft hat wollte ich es nun auch testen, aber ich scheitere natürlich an der Installation_ID. Alle anderen Werte habe ich. CURL ist ebenfalls in der Powershell am Start. Könnt ihr bitte kurz beschreiben, wie ich den Aufruf nun machen muss um die Installation_ID zu erhalten?

hbpv commented:
Hey

Für alle die keinen Zugriff auf Curl haben, weil Sie HA auf dem Pi oder x86 direkt installiert haben und das evcc Admin nutzen, ein möglicher workaround. Vielleicht schaffe ich es die Tage einen Nodered Code zu exportieren der alles enthält was man braucht.. 

https://www.rustimation.eu/index.php/1_zugang_api/

Das ist zwar eine gesamte Anbindung, aber ich denke daraus kann man sicher den Part der InstallationID nutzen. 

Vg

Cruiserlive commented:
Vermutlich ganz gut beschrieben von ChatGPT, teste es heute Abend mal:

🔧 Voraussetzungen

Du brauchst:
	•	curl (für HTTP-Anfragen)
	•	jq (zum Parsen von JSON)
	•	Ein Viessmann Entwicklerkonto: https://app.developer.viessmann.com/
	•	Client ID, Benutzername, Passwort

🧩 Vorbereitung
	1.	Registriere dich auf https://app.developer.viessmann.com/
	2.	Erstelle dort eine Anwendung und notiere dir die:
	•	Client ID
	•	(Redirect URI sollte sein: http://localhost:4200/)
	3.	Definiere in deinem Terminal die folgenden Umgebungsvariablen:

Bash:	
export VIESSMANN_USER="<dein Viessmann-Benutzername>"
export VIESSMANN_PASS="<dein Viessmann-Passwort>"
export VIESSMANN_CLIENT_ID="<deine Client ID>"

🚀 Terminal-Befehl zum Abrufen der Installations-ID

Führe folgenden kompletten Block im Terminal aus:

Bash:

VIESSMANN_REDIRECT_URI="http://localhost:4200/"
VIESSMANN_CODE_CHALLENGE="5M5nhkBfkWZCGfLZYcTL-l7esjPUN7PpZ4rq8k4cmys"
VIESSMANN_CODE_VERIFIER="6PygdmeK8JKPuuftlkc6q4ceyvjhMM_a_cJrPbcmcLc-SPjx2ZXTYr-SOofPUBydQ3McNYRy7Hibc2L2WtVLJFpOQ~Qbgic455ArKjUz9_UiTLnO6q8A3e.I_fIF8hAo"

curl -X POST "https://iam.viessmann.com/idp/v2/token" \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -d "grant_type=password&username=$VIESSMANN_USER&password=$VIESSMANN_PASS&client_id=$VIESSMANN_CLIENT_ID&scope=IoT%20User&redirect_uri=$VIESSMANN_REDIRECT_URI&code_challenge=$VIESSMANN_CODE_CHALLENGE&code_challenge_method=S256"

# Die Antwort enthält ein access_token – benutze das dann wie folgt:
# (Zuerst access_token aus Antwort extrahieren, z. B. manuell oder mit jq)

ACCESS_TOKEN="<dein Access Token>"

curl -X GET "https://api.viessmann.com/iot/v1/equipment/installations" \
  -H "Authorization: Bearer $ACCESS_TOKEN" | jq .

🔍 Was bekommst du?

Die Ausgabe enthält eine Liste mit deinen Installationen, und dort findest du:
	•	installationId
	•	gatewaySerial
	•	ggf. deviceId

Diese kannst du dann in deiner Konfiguration verwenden.

Die Codes von Viessmann könnten noch für private Zwecke funktionieren. Muss mir das heute Abend mal anschauen

MasterGoofy commented:
Als alter Windowsmensch habe ich versucht in der Eingabeaufforderung folgendes zu tun :
set VIESSMANN_USER="mastergoofy@gmx.de" #Ist ein anderer Account#
set VIESSMANN_PASS="Geheim1234" #Auch das Passwort ist ein anderes#
set VIESSMANN_CLIENT_ID="53456346345634566534653466" #Ebenso#
set VIESSMANN_REDIRECT_URI="http://localhost:4200/" #aus der Anleitung#
set VIESSMANN_CODE_CHALLENGE="5M5nhkBfkWZCGfLZYcTL-l7esjPUN7PpZ4rq8k4cmys" #aus der Anleitung#
set VIESSMANN_CODE_VERIFIER="6PygdmeK8JKPuuftlkc6q4ceyvjhMM_a_cJrPbcmcLc-SPjx2ZXTYr-SOofPUBydQ3McNYRy7Hibc2L2WtVLJFpOQ~Qbgic455ArKjUz9_UiTLnO6q8A3e.I_fIF8hAo" #aus der Anleitung#
curl -X POST "https://iam.viessmann.com/idp/v2/token" -H "Content-Type: application/x-www-form-urlencoded" -d "grant_type=password&username=$VIESSMANN_USER&password=$VIESSMANN_PASS&client_id=$VIESSMANN_CLIENT_ID&scope=IoT%20User&redirect_uri=$VIESSMANN_REDIRECT_URI&code_challenge=$VIESSMANN_CODE_CHALLENGE&code_challenge_method=S256"

Beim Absetzen des CURL Befehls kommt dann :
{"error":"invalid-token-request"}

mpollmeier commented:
Die Hilfe war noch nicht fertig, ich habe an der Stelle pausiert weil das Layout dann kaputt war. Ist aber keine große Sache, ich mache noch nen PR um das zu vervollständigen. 

Das Layout ist zwar nicht mehr zerstört, dafür wird aber der Text einfach abgeschnitten. Der User soll diesen ja komplett kopieren, das ist zwar möglich aber so nicht ideal. Naja, ist wohl eh eher eine "Profi" Integration, solange wir die InstallationID nicht automatisiert ermitteln. 
https://github.com/evcc-io/evcc/pull/20111#discussion_r2113991713

mpollmeier commented:
Für alle, die es schonmal testen wollen:

```
VIESSMANN_USER=<your-user>
VIESSMANN_PASS=<your-password>
VIESSMANN_CLIENT_ID=<your-clientid>
```

Then execute the following (n.b. it's best to paste the entire block as-is, since the intermediate 'CODE' is only valid for 20 seconds):

```
VIESSMANN_REDIRECT_URI="http://localhost:4200/"
VIESSMANN_CODE_CHALLENGE="5M5nhkBfkWZCGfLZYcTL-l7esjPUN7PpZ4rq8k4cmys"
VIESSMANN_CODE_VERIFIER="6PygdmeK8JKPuuftlkc6q4ceyvjhMM_a_cJrPbcmcLc-SPjx2ZXTYr-SOofPUBydQ3McNYRy7Hibc2L2WtVLJFpOQ~Qbgic455ArKjUz9_UiTLnO6q8A3e.I_fIF8hAo"

VIESSMANN_CODE=$(curl -X POST --silent \
  --user $VIESSMANN_USER:$VIESSMANN_PASS \
  --output /dev/null \
  --dump-header -    \
  "https://iam.viessmann.com/idp/v3/authorize?client_id=$VIESSMANN_CLIENT_ID&redirect_uri=$VIESSMANN_REDIRECT_URI&scope=IoT%20User%20offline_access&response_type=code&code_challenge=$VIESSMANN_CODE_CHALLENGE&code_challenge_method=S256" \
  | grep "^location: "            \
  | sed 's/.*\?code=\(.*\).*/\1/' \
  | tr -d '[:space:]')

TOKEN_RESPONSE=$(curl -XPOST --silent \
  -H "Content-Type: application/x-www-form-urlencoded" \
  --data "grant_type=authorization_code&client_id=$VIESSMANN_CLIENT_ID&redirect_uri=$VIESSMANN_REDIRECT_URI&code_verifier=$VIESSMANN_CODE_VERIFIER&code=$VIESSMANN_CODE" \
  https://iam.viessmann.com/idp/v3/token)

VIESSMANN_TOKEN=$(echo $TOKEN_RESPONSE | jq --raw-output .access_token)

curl -X GET https://api.viessmann.com/iot/v1/equipment/installations?includeGateways=true -H "Authorization: Bearer $VIESSMANN_TOKEN" > viessmann-installation.json
VIESSMANN_INSTALLATION_ID=$(jq '.data[].id' viessmann-installation.json)
VIESSMANN_GATEWAY_SERIAL=$(jq '.data[].gateways[].serial' --raw-output viessmann-installation.json)

MasterGoofy commented:
Dank einer Ubuntu VM habe ich die Scripte ausgeführt.
Aus der viessmann-installation.json habe ich die data.id in evcc in die Installation ID übernommen.
Klicke ich auf prüfen kommt :
Status: fehlgeschlagen
chargeStatus: dimm mode

mpollmeier commented:
Ich verstehe nicht ganz. Es geht vermutlich um diese hier? Aus weishaupt-wpm.yaml:
```
    - case: 1 # normal
    - case: 2 # boost
    - case: 3 # dimm
```

In viessmann.yaml fehlt demnach `case: 3 #dimm`, richtig?
Was bedeutet denn dimm? Wir brauchen doch auch nur zwei modes eigentlich, boost und normal...

mpollmeier commented:
Hier noch die restlichen Docs: https://github.com/evcc-io/evcc/pull/21568

@andig kannst du noch was zum dem `dimm` mode sagen, damit wir es ans Laufen bekommen?

andig commented:
Es hat sich nur die Numerierung der Modi geändert, siehe https://docs.evcc.io/docs/devices/heating. Wenn wir `dimm` nicht haben hilft 

        - source: error
          error: ErrNotAvailable

mpollmeier commented:
👍🏻  https://github.com/evcc-io/evcc/pull/21574

mpollmeier commented:
hmm, mit 448ef7db131a03ffdbf2ea1f5462c40e204d2985 bekomme ich immer noch denselben Fehler...
![image](https://github.com/user-attachments/assets/b179ceed-5726-4a3a-9593-a0553b136efe)

andig commented:
Kannst du das mal bitte mit

    evcc charger 

testen?

@naltatis der Text hinter `dimm mode` erscheint abgeschnitten?

andig commented:
@naltatis nvm. Die Fehlermeldung ist so kryptisch.

andig commented:
Eine Abfrage je Minute wird nicht genügen. Dafür müssten wir evcc- ggf. optional- abgewöhnen, den `Enabled()` Status nachzulesen. Das passiert sonst bei jedem Durchlauf da zumindest theoretisch auch andere Einheiten steuern könnten.

JanGew commented:
Ich habe gerade noch mal nachgesehen. Selbst bei den kostenpflichtigen Zugängen bei Viessmann sind es nur 3000 API calls pro Tag. Das wird also knapp... 😅

MasterGoofy commented:
Hallo, ich habe hier leider noch Probleme. 
Ich habe meine Wärmepumpe hinzugefügt. 
Sie wird als verbunden aber nicht Ladebereit angezeigt :
![Screenshot_20250608_151258_Chrome](https://github.com/user-attachments/assets/0d42578f-cd2d-450f-899b-b47f6f3755df)
Wenn ich nun manuell einen Aufheizvorgang starten möchte bekomme ich eine Fehlermeldung :
![Screenshot_20250613_105417_Home Assistant](https://github.com/user-attachments/assets/8db857f8-08ea-484c-b4ed-9b5f9ad7c24e)
Wo habe ich etwas falsch gemacht?

MasterGoofy commented:
Also wahnsinnig ergiebig ist der Logeintrag dazu auch nicht :
```
[db:9 ] ERROR 2025/06/15 11:20:30 unexpected status: 404 (Not Found)
```

andig commented:
> Mach bitte eine Discussion mit Logfile.

Für APIs braucht es immer ein trace log.

phL9r commented:
Hallo! Ich habe gestern folgende Mail bekommen, die Hinweise auf eine API Änderung gibt. Ist das schon bekannt / in der Umsetzung? Abschaltung der "alten" API läuft im Q4 2025.

[](https://viessmann.mailingmanager.online/letter/?20329_10991249_604EDFC9C1CFC89354FD6EA060331718)

mpollmeier commented:
https://github.com/evcc-io/evcc/pull/22350

MasterGoofy commented:
@hbpv : Kannst du bitte ein Bild oder Diagramm einstellen wie du die Shelly Plus (laut IDs setzt du 2 ein) mit der Wärmepumpe verbunden hast?

hbpv commented:
> @hbpv : Kannst du bitte ein Bild oder Diagramm einstellen wie du die Shelly Plus (laut IDs setzt du 2 ein) mit der Wärmepumpe verbunden hast?

Hi - sorry für die Verwirrung, das zweite Shelly stammt noch aus einer "Testphase" in der ich lieber erstmal eine Steckdose statt die Wärmepumpe geschalten habe um das ganze quasi erstmal zu simulieren.
Tatsächlich ist an der WP nur ein Shelly, welches den Modus siehe Anhang schaltet. Der Anschluss ist von der WP abhängig, es gibt da aber viele Einträge im Viessmann Forum. Ein Beispiel hier: https://community.viessmann.de/t5/Waermepumpe-Hybridsysteme/SG-Ready-bei-Vitocal-200-S-an-Klemme-X216-1-amp-4-und-X3-quot-L/td-p/104368

<img width="636" height="623" alt="SmartGrid 3 (1)" src="https://github.com/user-attachments/assets/fa830c23-48d1-4c95-945c-883a62d4da14" />
